### PR TITLE
Cleanup & Expansion

### PR DESCRIPTION
**Pull request by markhayden**
First step in actually closing: https://github.com/lytics/lio/issues/6881

Once the `delete-junk` branch gets thumbed / merged we can probably alter this pr to be against gh-pages but for now this will make it easier to review the code related changes. Also plan on rebasing this thing this afternoon.

`out` are the compiled files so all comments should be against files in `src`

## Cleanup
Migrated everything to use gulp.

## Tests
Added gulp, jasmine, karma, phantom for unit tests. Wrote tests for everything I could think of. I am sure ther are some missing but here is a shortlist of the things that are covered:

* extend function
* setcookie
* getcookie
* async queue
* send
* identify
* pageview
* mock
* block
* etc

## New Functionality
In support of 6881 I introduced the concept of a `mock` and a `hold` on sending events. We needed the ability to basically pause all sends and then build an object of that send queued data so that we could pass it to our `/me` endpoint and get the most current state of the entity. This will make strides in facilitating the `first visit` use case as well as many other use cases negatively effected by latency.

In a nutshell it routes all send events to a queue. Then when send the mock event it pulls all that data out of the queue, extends the current payload and does all the processing up the point where it would actually be sent. It then returns the payload via callback so we can handle it in `/tag/lio.js`.

When block is lifted all existing events get processed normally.

### TODO:
* Add Readme
* Rebase all Commits

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,2 +1,5 @@
+.DS_Store
 node_modules
-.env.json
\ No newline at end of file
+.env
+.env.json
```
Comment (DingoEatingFuzz) [inline .gitignore]: An env and an env.json?

File: LICENSE [hunk 2]
```diff
@@ -1,4 +1,4 @@
-Copyright (c) 2012 Lytics.io
+Copyright (c) 2016 Lytics.io
```
Comment (DingoEatingFuzz) [inline LICENSE]: lol

Comment (araddon) [inline LICENSE]: lol

File: out/async.js [hunk 3]
```diff
@@ -1,53 +1,65 @@
...
+          _c: {},
+          ts: (new Date()).getTime()
+      },
+      l = false,
+      w = window,
+      d = document,
+      src = "/static/io",
+      ext = ".min.js",
+      as = Array.prototype.slice,
+      js = "//api.lytics.io" || "//c.lytics.io",
```
Comment (araddon) [inline out/async.js]: can we just use c.?

File: gulpfile.js [hunk 4]
```diff
@@ -0,0 +1,190 @@
+var gulp = require('gulp'),
+    uglify = require('gulp-uglify'),
+    rename = require('gulp-rename'),
+    connect = require('gulp-connect'),
+    replace = require('gulp-replace'),
+    env = require('gulp-env'),
+    Server = require('karma').Server,
+    open = require('gulp-open'),
+    karma = require('karma'),
```
Comment (DingoEatingFuzz) [inline gulpfile.js]: You can reorder this to not require karma twice.

```js
karma = require('karma'),
Server = karma.Server,
open = require('gulp-open')
```

File: src/io.js [hunk 5]
```diff
@@ -498,7 +531,7 @@
       if (jstag.config.qsargs && isArray(jstag.config.qsargs)) {
         var qsa = null
         for (var i = jstag.config.qsargs.length - 1; i >= 0; i--) {
-          qsa = jstag.config.qsargs[i]
+          qsa = jstag.config.qsarg
```
Comment (araddon) [inline src/io.js]: ?? this doesn't seem right?

File: src/outlio.js [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    _c: {},
+    ts: (new Date()).getTime()
+  },
+  l = false,
+  w = window,
+  d = document,
+  src = "/static/io",
+  ext = ".min.js",
+  as = Array.prototype.slice,
+  js = "//api.lytics.io" || "//c.lytics.io",
```
Comment (araddon) [inline src/outlio.js]: can we do just c.?

File: gulpfile.js [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+  master_url = production_url;
+}
+
+/*
+* sets master version information
+*/
+var setVersion = function(){
+  var obj = JSON.parse(fs.readFileSync('src/versioning.json', 'utf8'));
+
+  console.log(obj);
```
Comment (DingoEatingFuzz) [inline gulpfile.js]: I don't think you want this.

File: LICENSE (hunk 2 above)
Comment (humanchimp) [inline LICENSE]: Shouldn't it say `Copyright (c) 2016 Lytics, Inc.`?

File: gulpfile.js [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+gulp.task('build:legacy', function (done) {
+  gulp.src(['src/legacy/async.js', 'src/legacy/io.js'])
+    .pipe(gulp.dest('out/'))
+      .pipe(uglify())
+      .pipe(rename({
+          suffix: '.min'
+      }))
+      .pipe(gulp.dest('out/'))
+      done();
+});
```
Comment (DingoEatingFuzz) [inline gulpfile.js]: The indentation here is weird. It looks like the second pipe is nested in within the first one, but it isn't.

Also, there is no semi-colon after `.pipe(gulp.dest('out/'))` and no dot before `done()`. Which is weird formatting and also makes me wonder if it's a bug.

File: gulpfile.js [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+    .pipe(replace('{{asyncversion}}', asyncversion))
+    .pipe(replace('{{ioversion}}', asyncversion))
+    .pipe(replace('{{initobj}}', JSON.stringify(initobj, null, 2)))
+    .pipe(gulp.dest('out/'+version))
+      .pipe(uglify())
+      .pipe(rename({
+          suffix: '.min'
+      }))
+      .pipe(gulp.dest('out/'+version))
+      done();
```
Comment (DingoEatingFuzz) [inline gulpfile.js]: More weird formatting.

File: gulpfile.js [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+});
+
+gulp.task('build:development', function (done) {
+  var initobj = generateConfig('development');
+
+	gulp.src(['src/async.js', 'src/io.js'])
```
Comment (DingoEatingFuzz) [inline gulpfile.js]: ...a tab? Where did that come from?

File: gulpfile.js [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+    .pipe(replace('{{version}}', version))
+    .pipe(replace('{{asyncversion}}', asyncversion))
+    .pipe(replace('{{ioversion}}', ioversion))
+    .pipe(replace('{{initobj}}', JSON.stringify(initobj, null, 2)))
+		.pipe(gulp.dest('out/'+version))
+    	.pipe(uglify())
+    	.pipe(rename({
+      		suffix: '.min'
```
Comment (DingoEatingFuzz) [inline gulpfile.js]: A tab AND spaces? @markhayden, please.

File: gulpfile.js [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+//   }).start();
+// });
+
+// gulp.task('coverage', gulp.series('unit:coverage'), function() {
+//   return gulp.src('./coverage/index.html')
+//     .pipe(open());
+// });
+
+// builds for the development environment and runs all tests
+gulp.task('test', gulp.series('fixtures:test', 'build:production', 'build:legacy', 'asynctest', 'iotest', 'dualsendtest'));
```
Comment (DingoEatingFuzz) [inline gulpfile.js]: This is a pretty clunky way to run tests, and it's unfortunate that the server needs to be setup and torn down for each test suite.

Not a blocker though. Just a thing to improve in the future.

File: .gitignore (hunk 1 above)
Comment (markhayden) [inline .gitignore]: lol. fixed.

File: gulpfile.js (hunk 12 above)
Comment (markhayden) [inline gulpfile.js]: definitely can / should be improved. had planned on making a ticket to improve tests / re-integrate coverage and stuff after this thing gets merged.

File: out/async.js (hunk 3 above)
Comment (markhayden) [inline out/async.js]: yah definitely.

File: gulpfile.js (hunk 4 above)
Comment (markhayden) [inline gulpfile.js]: forgot about that one. can actually just remove `Server` and use `karma.Server` below i believe.

File: src/io.js (hunk 5 above)
Comment (markhayden) [inline src/io.js]: yah not sure what happened there. good catch, went back to how it was.

File: gulpfile.js (hunk 7 above)
Comment (markhayden) [inline gulpfile.js]: nope. fixed.

File: gulpfile.js (hunk 8 above)
Comment (markhayden) [inline gulpfile.js]: fixed

File: gulpfile.js (hunk 9 above)
Comment (markhayden) [inline gulpfile.js]: yah yah. fixed.

File: gulpfile.js (hunk 10 above)
Comment (markhayden) [inline gulpfile.js]: nobody knows! fixed.

File: gulpfile.js (hunk 11 above)
Comment (markhayden) [inline gulpfile.js]: i know how to keyboard. fixed.

File: out/v2/io.js [hunk 13]
```diff
@@ -0,0 +1,1091 @@
...
+          cb = arguments[i];
+          break;
+        case 'object' || 'array':
+          data = arguments[i];
+          break;
+        default:
+          console.warn("unable to process jstag.send event: unknown value type (" + typeof arguments[i] + ")");
+          break;
+      }
+    }
```
Comment (DingoEatingFuzz) [inline out/v2/io.js]: Um. I don't like this one bit. This is because of the signature difference between `send` and `identify`, right?

File: out/v2/io.js [hunk 14]
```diff
@@ -0,0 +1,1091 @@
...
+        sendHandler(jstag.config.payloadQueue[i]);
+      }
+
+      jstag.config.payloadQueue = [];
+    }else{
+      // invalid status
+      throw "invalid jstag.block status requested";
+    }
+  }
+  jstag['block'] = block;
```
Comment (DingoEatingFuzz) [inline out/v2/io.js]: I don't like the ergonomics of this function.

Instead of something like this:

```js
jstag.block(true);
// do some things before sending things
jstag.block(false);
```

How about

```js
jstag.block();
// do some things before sending things
jstag.unblock();
```

Or it could use a block pattern, like file io generally does.

```js
jstag.pause(function() {
  // do some things before sending things,
});
```

```js
jstag.pause(function(optionalResume) {
  somethingAsync.then(optionalResume);
});
```

File: out/v2/io.js [hunk 15]
```diff
@@ -0,0 +1,1091 @@
...
+      jstag.config.blockload = true;
+    }else if(status===false){
+      // status changed to false, process the payloadQueue
+      jstag.config.blockload = false;
+
+      for (var i = 0; i < jstag.config.payloadQueue.length; i++) {
+        sendHandler(jstag.config.payloadQueue[i]);
+      }
+
+      jstag.config.payloadQueue = [];
```
Comment (DingoEatingFuzz) [inline out/v2/io.js]: It'll probably never come up, but since this is setting `payloadQueue` to a new array, if someone is holding on to a reference to `payloadQueue`, this will break that reference.

You can instead do `jstag.config.payloadQueue.length = 0`, which will empty the existing array rather than creating a new one.

**Comment by araddon** [comment]
looks good, probably would like to see the `identify` fix, use `identify` stream (if none supplied)?

:+1:

File: out/v2/io.js [hunk 16]
```diff
@@ -0,0 +1,1091 @@
...
+      jstag.config.payloadQueue.push(payload);
+      return;
+    }
+
+    if(payload.mock){
+      var tempqdata = {};
+
+      for (var mq = 0, mqlen = jstag.config.payloadQueue.length; mq < mqlen; mq++){
+        tempqdata = extend(tempqdata, jstag.config.payloadQueue[mq].data);
+      }
```
Comment (DingoEatingFuzz) [inline out/v2/io.js]: What is the purpose of extending the payload with everything in the queue?

File: out/v2/io.js [hunk 17]
```diff
@@ -0,0 +1,1091 @@
...
+        case 'string':
+          stream = arguments[i];
+          break;
+        case 'boolean':
+          mock = arguments[i];
+          break;
+        case 'function':
+          cb = arguments[i];
+          break;
+        case 'object' || 'array':
```
Comment (humanchimp) [inline out/v2/io.js]: this seems like a mistake to me. It will greedily evaluate `'object' || 'array'` which will always evaluate to "object". Also, there is no `typeof` in JavaScript that will return array. Is it for documentation purposes?

File: out/v2/io.js [hunk 18]
```diff
@@ -0,0 +1,1091 @@
...
+   * public function wraps send function and adds mock param by default
+   * @param stream:  (string) optional name of stream to send to
+   * @param data:  the javascript object to be sent
+   * @param callback (optional):  the function to be called upon triggering elsewhere.
+   * @param mock (optional):  should the data being processed be mocked or sent through to lytics
+  */
+  function mock(){
+    // when mocking we pass true as the final argument to prevent sending data
+    [].push.call(arguments, true);
+    jstag.send.apply(this, arguments);
```
Comment (DingoEatingFuzz) [inline out/v2/io.js]: Pushing onto the arguments object is code smelly. It would be better to copy arguments, then push to a proper array, then send that.

```
var args = [].slice.call(arguments);
args.push(true);
jstag.send.apply(jstag, args);
```

File: out/v2/io.js (hunk 16 above)
Comment (markhayden) [inline out/v2/io.js]: my assumption is that we'd want to incorporate as much data as possible into the actual state call. this would incorporate anything that was sent before our tag was loaded. likewise it would allow for us to pull in things like a custom data layer or gtm dataLayer prior to hitting the /me endpoint so we'd get the fullest profile possible.

File: out/v2/io.js (hunk 15 above)
Comment (markhayden) [inline out/v2/io.js]: good to know, fixed.

File: out/v2/io.js [hunk 19]
```diff
@@ -0,0 +1,1091 @@
...
+  //  @data = object of key:value properties to collect
+  function identify(){
+    var payload;
+
+    // handle the event parsing
+    payload = parseEvent.apply(this, arguments);
+
+    // force overriding the stream and user_id seem pretty terrible but will address in another pr
+    // force stream to be default
+    payload.steram = "default";
```
Comment (DingoEatingFuzz) [inline out/v2/io.js]: Forcing a stream change seems showstoppingly bad, considering there are instances of jstag in the wild that are specifically set up to never send to `default`.

Fortunately, this won't override the stream because you spelled it wrong :trollface:

File: out/v2/io.js (hunk 13 above)
Comment (markhayden) [inline out/v2/io.js]: partly due to identify. also in part due to the lack of requiring a stream. essentially all of these are currently valid so we need to handle them:

```
jstag.send("stream")
jstag.send("stream", {})
jstag.send({})
jstag.send({}, callback)
```
i agree its not...likeable? but not sure of a better alternative that doesn't risk messing up all our existing jstag.send calls in place :S

File: src/versioning.json [hunk 20]
```diff
@@ -0,0 +1,5 @@
+{
+  "version": "v2",
+  "asyncversion": "2.0.0",
+  "ioversion": "2.0.0"
```
Comment (DingoEatingFuzz) [inline src/versioning.json]: What is the point of `version`?

File: out/v2/io.js (hunk 14 above)
Comment (markhayden) [inline out/v2/io.js]: went with block and unblock, good feedback

Comment (markhayden) [inline out/v2/io.js]: and my tests still pass. yes, i said tests!

File: out/v2/io.js (hunk 13 above)
Comment (markhayden) [inline out/v2/io.js]: image

File: out/v2/io.js (hunk 17 above)
Comment (markhayden) [inline out/v2/io.js]: yah might not be necessary / could be gross. mainly for documentation so that we remember / whomever rewrites this from the ground up knows that the payload can be {} or [].

File: tests/coreIoSpec.js [hunk 21]
```diff
@@ -0,0 +1,349 @@
...
+  obj = {"test":"one"};
+
+  it("should create the correct payload from a wide variety of send/mock events", function() {
+    // no object : invalid call
+    resp = jstag.parseEvent();
+    expect(resp).toEqual({ data: {}, callback: undefined, stream: "default", mock: false });
+
+    // just a stream and no object : invalid call
+    resp = jstag.parseEvent();
+    expect(jstag.parseEvent("fakestream")).toEqual({ data: {}, callback: undefined, stream: "fakestream", mock: false });
```
Comment (DingoEatingFuzz) [inline tests/coreIoSpec.js]: You are setting resp but then not using it.

File: out/v2/io.js (hunk 19 above)
Comment (markhayden) [inline out/v2/io.js]: lol. the troll is real. this was a thing that is super gross and i kind of just ignored it. got the go ahead to remove / revamp it tho so going to go ahead and do that. and push an update.

File: out/v2/io.js (hunk 18 above)
Comment (markhayden) [inline out/v2/io.js]: applied body wash. removed smell.

File: src/versioning.json (hunk 20 above)
Comment (markhayden) [inline src/versioning.json]: folder name? lol?

File: tests/coreIoSpec.js (hunk 21 above)
Comment (markhayden) [inline tests/coreIoSpec.js]: copy pasta leftovers for that edge case, removed.

File: tests/coreIoSpec.js [hunk 22]
```diff
@@ -0,0 +1,349 @@
...
+      // first send should not include uid, but does include sesstart
+      expect(async1.dataMsg).toEqual('one=one&_ts=' + async1.data._ts + '&_nmob=t&_device=desktop&url=localhost%3A9876%2Fcontext.html&_if=t&_v='+window.__karma__.config.ioversion+'&_e=pv&_sesstart=1&_tz=-7&_ul=en-US&_sz=1024x768&_uid=' + async1.data._uid + '&_getid=t&_ca=jstag1');
+
+      // future sends should not include session but do include new _uid
+      expect(async2.dataMsg).toEqual('two=one&_ts=' + async2.data._ts + '&_nmob=t&_device=desktop&url=localhost%3A9876%2Fcontext.html&_if=t&_uid=' + async2.data._uid + '&_getid=t&_v='+window.__karma__.config.ioversion+'&_ca=jstag1');
+      expect(async3.dataMsg).toEqual('three=one&_ts=' + async3.data._ts + '&_nmob=t&_device=desktop&url=localhost%3A9876%2Fcontext.html&_if=t&_uid=' + async3.data._uid + '&_getid=t&_v='+window.__karma__.config.ioversion+'&_ca=jstag1');
+      expect(async4.dataMsg).toEqual('four=one&_ts=' + async4.data._ts + '&_nmob=t&_device=desktop&url=localhost%3A9876%2Fcontext.html&_if=t&_uid=' + async4.data._uid + '&_getid=t&_v='+window.__karma__.config.ioversion+'&_ca=jstag1');
```
Comment (DingoEatingFuzz) [inline tests/coreIoSpec.js]: These are some brutal lines, but we knew that things were gonna get nasty since we are testing environment things.

File: tests/coreIoSpec.js [hunk 23]
```diff
@@ -0,0 +1,349 @@
...
+  });
+});
+
+describe("testing that jstag.send can handle an array in addition to an object", function () {
+  var async1;
+
+  function testSend(done) {
+    jstag.send("teststream", [{"one":"one"}, {"two":"two"}], function(opts, self){
```
Comment (DingoEatingFuzz) [inline tests/coreIoSpec.js]: It's generally bad to have keys and values be the same. It makes it impossible to detect if the key is going where the value should be and vice versa.

File: tests/coreIoSpec.js [hunk 24]
```diff
@@ -0,0 +1,349 @@
...
+describe("verify that we can properly process a pageview event", function () {
+  var async1, count=0;
+
+  function testPageView(done) {
+    jstag.pageView("teststream", function(opts, self){
+      async1 = opts;
+      count++;
+      if(count >= 1){
+        done();
+      }
```
Comment (DingoEatingFuzz) [inline tests/coreIoSpec.js]: No need to deal with tracking `count` if there is only one test.

File: tests/coreIoSpec.js [hunk 25]
```diff
@@ -0,0 +1,349 @@
...
+    });
+  }
+
+  beforeEach(function (done) {
+      testPageView(done);
+  });
+
+  it("should add and/or alter the _e param and start session", function () {
+    expect(async1.dataMsg).toEqual('_e=pv&_sesstart=1&_tz=-7&_ul=en-US&_sz=1024x768&_ts=' + async1.data._ts + '&_nmob=t&_device=desktop&url=localhost%3A9876%2Fcontext.html&_if=t&_uid=' + async1.data._uid + '&_getid=t&_v='+window.__karma__.config.ioversion+'&_ca=jstag1');
+  });
```
Comment (DingoEatingFuzz) [inline tests/coreIoSpec.js]: Not worth dealing with now, but eventually we should come up with better helpers for these kinds of tests. Both to improve readability (since it's incredibly difficult to figure out why values are different when they are different) and to allow for values to come in in arbitrary order, since order doesn't (shouldn't?) affect behavior.

File: tests/coreIoSpec.js [hunk 26]
```diff
@@ -0,0 +1,349 @@
...
+
+  console.llog
```
Comment (DingoEatingFuzz) [inline tests/coreIoSpec.js]: ...

File: tests/coreIoSpec.js [hunk 27]
```diff
@@ -0,0 +1,349 @@
...
+});
+
+describe("verify that we can properly process an identify event", function () {
+  var async1, count=0;
+
+  function testIdentify(done) {
+    jstag.identify("myfakeuserid", function(opts, self){
```
Comment (DingoEatingFuzz) [inline tests/coreIoSpec.js]: There's no need to declare `self` in the arguments if you aren't going to use it.

File: tests/coreIoSpec.js [hunk 28]
```diff
@@ -0,0 +1,349 @@
...
+
+    jstag.send("streamname", {"test":"test"}, function(opts, self){
+      async1 = opts;
+      count++;
+      if(count >= 2){
+        done();
+      }
+    });
+
+    jstag.send("streamname", {"test2":"test2"}, function(opts, self){
```
Comment (DingoEatingFuzz) [inline tests/coreIoSpec.js]: More instances of keys and values matching.

File: tests/coreIoSpec.js [hunk 29]
```diff
@@ -0,0 +1,349 @@
...
+      count++;
+      if(count >= 2){
+        done();
+      }
+    });
+
+    expect(jstag.config.blockload).toBe(true);
+    expect(jstag.config.payloadQueue.length).toEqual(2);
+
+    jstag.block(false);
```
Comment (DingoEatingFuzz) [inline tests/coreIoSpec.js]: It would be good to get this line and any other state reseting code in a global `beforeEach` to make sure tests don't poison other tests.

File: tests/coreIoSpec.js [hunk 30]
```diff
@@ -0,0 +1,349 @@
...
+  beforeEach(function (done) {
+    testSend(done);
+  });
+
+  it("should merge the payload from the two blocked events and one mock event", function() {
+    expect(async1.dataMsg).toEqual('test=test&test2=test2&test3=test3&_ts=' + async1.data._ts + '&_nmob=t&_device=desktop&url=localhost%3A9876%2Fcontext.html&_if=t&_uid=' + async1.data._uid + '&_getid=t&_v='+window.__karma__.config.ioversion+'&_ca=jstag1');
+  });
+});
+
+describe("verify that we can get and set cookies as well as that the core cookies are set (seerid and seerses)", function() {
```
Comment (DingoEatingFuzz) [inline tests/coreIoSpec.js]: The test name doesn't read well.

Comment (markhayden) [inline tests/coreIoSpec.js]: fixed

File: tests/coreIoSpec.js (hunk 28 above)
Comment (markhayden) [inline tests/coreIoSpec.js]: fixed

File: tests/coreIoSpec.js (hunk 27 above)
Comment (markhayden) [inline tests/coreIoSpec.js]: fixed

File: tests/coreIoSpec.js (hunk 26 above)
Comment (markhayden) [inline tests/coreIoSpec.js]: ...

File: tests/coreIoSpec.js (hunk 25 above)
Comment (markhayden) [inline tests/coreIoSpec.js]: agreed....huge pain. will leave for now then double back.

File: tests/coreIoSpec.js (hunk 24 above)
Comment (markhayden) [inline tests/coreIoSpec.js]: fixed

File: tests/coreIoSpec.js (hunk 23 above)
Comment (markhayden) [inline tests/coreIoSpec.js]: fixed across the board

File: tests/coreIoSpec.js (hunk 22 above)
Comment (markhayden) [inline tests/coreIoSpec.js]: yah i think we could at least make a helper or something that does the replacement for you so its not quite so brutal or something after we get this pr out.

File: tests/coreIoSpec.js (hunk 29 above)
Comment (markhayden) [inline tests/coreIoSpec.js]: agreed. this was a pain to test because of the callbacks and such. right now im 98% sure nothing is polluting. wondering if we make this another ticket and fix after merge just to keep things more organized / its likely to be a pretty decent sized change that will effect all the output testing.

File: out/v2/io.js (hunk 16 above)
Comment (DingoEatingFuzz) [inline out/v2/io.js]: But wouldn't this also result in sending duplicate data, since we call `sendHandler` for every item in the queue when we flush it on `unblock`?

File: src/async.js [hunk 31]
```diff
@@ -0,0 +1,64 @@
...
+    t._c.blockload = true;
+    return this;
+  };
+  t.unblock = function() {
+    t._c.blockload = false;
+    return this;
+  };
+  return t;
+}(),
+window.jstag.init({{initobj}});
```
Comment (DingoEatingFuzz) [inline src/async.js]: Do we want to make this file as small as possible right now to avoid having to update all the tag managers yet again?

If so, there are some clever things we can do to reduce boilerplate in here.

File: src/io.js [hunk 32]
```diff
@@ -0,0 +1,1109 @@
...
+          break;
+        case 'boolean':
+          mock = arguments[i];
+          break;
+        case 'function':
+          cb = arguments[i];
+          break;
+        case 'object' || 'array':
+          data = arguments[i];
+          break;
```
Comment (DingoEatingFuzz) [inline src/io.js]: To reiterate what @humanchimp was saying, even if you want to document `array` in here for readability, this case statement is suspect. This isn't an expression for case 'object' or case 'array', this is an expression for a case of the immediate result of the expression `'object' || 'array'`, which always be `'object'` because it's truthy. Which means if `typeof arguments[i]` returned `'array'` (which isn't possible, but humor me) this would fall through to the default case.

What I believe you actually want is cascading cases. So if `typeof arguments[i]` is either `'object'` or `'array'`, the result is setting `data` to `arguments[i]` instead of warning that we couldn't process the event.

```js
case 'object':
case 'array':
  data = arguments[i];
  break;
```

Again, calling `typeof []` results in `'object'`, so this whole thing is kinda moot.

File: out/v2/async.js [hunk 33]
```diff
@@ -0,0 +1,69 @@
...
+    return this;
+  };
+  return t;
+}(),
+window.jstag.init({
+  "cid": "12",
+  "url": "//marksthingy.ngrok.io",
+  "loadid": false,
+  "min": false
+});
```
Comment (DingoEatingFuzz) [inline out/v2/async.js]: Testing something?

Comment (markhayden) [inline out/v2/async.js]: nope thats the new url.

it'll be fixed here in a sec.

File: src/io.js (hunk 32 above)
Comment (markhayden) [inline src/io.js]: just removed and added a comment :)

**Comment by markhayden** [comment]
have to re-open this pr as well against develop. will ref so we have history but new pr should be clean and ready for one last pass by @araddon and @DingoEatingFuzz
